src/providers/shapeshift/shapeshift.ts

```diff
@@ -162,13 +194,241 @@ export class ShapeshiftProvider {
+      }
+
+      return JSON.stringify(errData);
+    } catch (e) {
+      this.logger.error(e);
+      return e;
+    }
+  }
+
+  public init = _.throttle(cb => {
+    if (_.isEmpty(this.credentials.CLIENT_ID)) {
+      return cb('ShapeShift is Disabled. Missing credentials.');
+    }
+    this.logger.debug('Trying to initialize ShapeShift...');
+
+    this.getStoredToken(accessToken => {
+      if (!accessToken) {
+        this.logger.debug('ShapeShift not linked');
+        return cb();
+      }
+      this.logger.debug('ShapeShift already has Token.');
+      this._getMainAccountId(accessToken, (err, accountId) => {
+        if (err) {
+          if (!err.errors) return cb(err);
+          if (err.errors && !_.isArray(err.errors)) return cb(err);
+
+          let expiredToken;
+          for (let i = 0; i < err.errors.length; i++) {
+            if (err.errors[i].id == 'expired_token') expiredToken = true;
+          }
+
+          if (expiredToken) {
+            this.logger.debug('Token Expired. Refresh and get new Token.');
+            this.persistenceProvider
+              .getShapeshiftRefreshToken(this.credentials.NETWORK)
+              .then(refreshToken => {
+                this._refreshToken(refreshToken, (err, newToken) => {
+                  if (err) return cb(err);
+                  this._getMainAccountId(newToken, (err, accountId) => {
+                    if (err) return cb(err);
+                    return cb(null, {
+                      accessToken: newToken,
+                      accountId
+                    });
+                  });
+                });
+              })
+              .catch(err => {
+                return cb(err);
+              });
+          } else {
+            return cb(err);
+          }
+        } else {
+          return cb(null, {
+            accessToken,
+            accountId
+          });
+        }
+      });
+    });
+  }, 10000);
+
+  private _getMainAccountId(accessToken, cb) {
+    this.getAccount(accessToken, (err, a) => {
+      if (err) {
+        this.logout();
+        return cb(err);
+      }
+      if (a.data.verificationStatus == 'NONE') {
+        return cb('unverified_account');
+      }
+      return cb(null, a.data.id);
+    });
+  }
+
+  public getAccount(token, cb) {
+    if (!token) return cb('Invalid Token');
+
+    let url = this.credentials.HOST + '/api/v1/users/me';
+    let headers = {
+      'Content-Type': 'application/json',
+      Accept: 'application/json',
+      Authorization: 'Bearer ' + token
+    };
+
+    this.http.get(url, { headers }).subscribe(
+      data => {
+        this.logger.info('ShapeShift: Get Account SUCCESS');
+        return cb(null, data);
+      },
+      data => {
+        this.logger.error(
+          'ShapeShift: Get Account ERROR ' +
+            data.status +
+            '. ' +
+            this.getErrorsAsString(data.error)
+        );
+        return cb(data.error);
+      }
+    );
+  }
+
+  public logout() {
```

Comment:
Logout should implement correct ShapeShift POST to revoke token

Reply:
This method only remove token from the storage but we need to revoke the token from ShapeShift in order to avoid someone to reuse it.